pkg/component/worker/utils.go

```diff
@@ -17,62 +17,154 @@ limitations under the License.
-	}
+	if joinToken == "" {
+		bootstrapKubeconfigPath = filepath.Join(k0sVars.DataDir, "kubelet-bootstrap.conf")
+		if !file.Exists(bootstrapKubeconfigPath) {
+			return fmt.Errorf("neiter regular nor bootstrap kubelet kubeconfig files exist and no join token given; dunno how to make kubelet authenticate to API server")
+		}
 
-	if tokenType := token.GetTokenType(clientCfg); tokenType != "kubelet-bootstrap" {
-		return fmt.Errorf("wrong token type %s, expected type: kubelet-bootstrap", tokenType)
+		var err error
+		bootstrapKubeconfig, err = clientcmd.LoadFromFile(bootstrapKubeconfigPath)
+		if err != nil {
+			return fmt.Errorf("failed to parse kubelet bootstrap kubeconfig from file: %w", err)
+		}
+	} else {
+		kubeconfig, err := token.DecodeJoinToken(joinToken)
+		if err != nil {
+			return fmt.Errorf("failed to decode join token: %w", err)
+		}
+
+		// Load the bootstrap kubeconfig to validate it
+		bootstrapKubeconfig, err = clientcmd.Load(kubeconfig)
+		if err != nil {
+			return fmt.Errorf("failed to parse kubelet bootstrap kubeconfig from join token: %w", err)
+		}
+
+		bootstrapKubeconfigPath, err = writeKubeletBootstrapKubeconfig(kubeconfig)
+		if err != nil {
+			return fmt.Errorf("failed to write kubelet bootstrap kubeconfig: %w", err)
+		}
+		defer func() {
+			if err := os.Remove(bootstrapKubeconfigPath); err != nil && !os.IsNotExist(err) {
+				logrus.WithError(err).Error("Failed to remove kubelet bootstrap kubeconfig file")
+			}
+		}()
+
+		logrus.Debug("Wrote kubelet bootstrap kubeconfig file: ", bootstrapKubeconfigPath)
 	}
 
 	kubeletCAPath := path.Join(k0sVars.CertRootDir, "ca.crt")
 	if !file.Exists(kubeletCAPath) {
 		if err := dir.Init(k0sVars.CertRootDir, constant.CertRootDirMode); err != nil {
 			return fmt.Errorf("failed to initialize directory '%s': %w", k0sVars.CertRootDir, err)
 		}
-		err = file.WriteContentAtomically(kubeletCAPath, clientCfg.Clusters["k0s"].CertificateAuthorityData, constant.CertMode)
+		err := file.WriteContentAtomically(kubeletCAPath, bootstrapKubeconfig.Clusters["k0s"].CertificateAuthorityData, constant.CertMode)
 		if err != nil {
 			return fmt.Errorf("failed to write ca client cert: %w", err)
 		}
 	}
-	err = file.WriteContentAtomically(k0sVars.KubeletBootstrapConfigPath, kubeconfig, constant.CertSecureMode)
+
+	if tokenType := token.GetTokenType(bootstrapKubeconfig); tokenType != "kubelet-bootstrap" {
+		return fmt.Errorf("wrong token type %s, expected type: kubelet-bootstrap", tokenType)
+	}
+
+	certDir := filepath.Join(k0sVars.DataDir, "kubelet", "pki")
+	if err := dir.Init(certDir, constant.DataDirMode); err != nil {
+		return fmt.Errorf("failed to initialize kubelet certificate directory: %w", err)
+	}
+
+	// FIXME: This is not completely on par with kubelet code. The kubelet has
+	// some means of retrieving the node name via the cloud provider, if one is
+	// configured.
+	nodeName, err := nodeutil.GetHostname("")
```

Comment:
Do we want to retrieve the hostname override somehow? Not sure how important that might be. Seems doable, but would require quite some code shuffling, i.e. we'd need to pull out the kubelet config part out of `kubelet.go` in order to load the worker profile early.

As far as I can tell, the node name here is only used as suffix to the user name in the certs, so that one can see which node authenticated with the API server.

Reply:
> As far as I can tell, the node name here is only used as suffix to the user name in the certs, so that one can see which node authenticated with the API server.

Does the token -> cert exchange already create the node object in the API? If it does, then we need to worry about the hostname override. If not then I think we do not have to worry about that.

Reply:
The bootstrapping won't create any node objects. But there's still something that requires that the cert's node name and the kubelet's node name match. I tried to join a worker via `k0s worker --kubelet-extra-args='--hostname-override=foobar'`. The kubelet won't be able to make progress and spams a lot of "Error getting node" err="node \"foobar\" not found" in the logs. Not sure why, but if the host names match, everything is well.

Maybe this is related to how the [Node authorization](https://kubernetes.io/docs/reference/access-authn-authz/node/) works (although the error message doesn't indicate any authz problems?). The docs there say:

> The value of `<nodeName>` **must** match precisely the name of the node as registered by the kubelet. By default, this is the host name as provided by `hostname`, or overridden via the [kubelet option](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) `--hostname-override`. However, when using the `--cloud-provider` kubelet option, the specific hostname may be determined by the cloud provider, ignoring the local `hostname` and the `--hostname-override` option. For specifics about how the kubelet determines the hostname, see the [kubelet options reference](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/).

So I guess k0s needs to make sure as best as possible that the node names match. I realized that the hostname override solely comes from k0s args, not from any dynamic source, so it's much more straight forward to use that setting also during bootstrapping. Will add that to the PR, along with an inttest that covers this case.

I'm not sure about the cloud provider stuff. In that case, kubelet will apparently _always_ use the node name from the cloud provider, even if the hostname override option is given. Assuming that kubelet will also fail if the names don't match, as in the non-cloud provider case, the only way to make k0s work with a cloud provider is to specify the right hostname override manually, so that k0s picks it up (even if kubelet will ignore it).

Reply:
I double-checked the kubelet code. K0s only supports external cloud providers, which actually seems to be a [dead code path](https://github.com/kubernetes/kubernetes/blob/v1.25.4/cmd/kubelet/app/server.go#L547-L559) in kubelet anyways. So I don't see any way k0s could come to a different conclusion about the node name than kubelet.